packages/contract/lib/reason.js

```diff
@@ -13,7 +13,13 @@ const reason = {
   _extract: function(res, web3, _interfaceAdapter) {
     //I'm not sure why interfaceAdapter is here if it's not used,
     //so I just put an underscore in front of its name for now...
-    if (!res || (!res.error && !res.result)) return;
+    if (!res) return;
+    if (!res.error && (
```

Comment:
This logic doesn't look correct to me; and if it somehow is correct, then I'm pretty sure it's been written in a way that's overcomplicated.

This conditional is trying to detect if there's *no* error, right?  So it ought to be checking that there is *both* no 2.x error *and* no 7.x error.  But instead it uses an `||`.  Moreover, it first checks `!res.error`, and then checks `res.error`... that can't be right.  (And if it *is* right, it probably ought to be able to use some Boolean algebra to simplify that.)

Reply:
Let's undo this change and instead borrow @haltman-at's change in https://github.com/trufflesuite/truffle/pull/4506.

Reply:
Or maybe we don't need to borrow #4506 (assuming tests pass :)